File: 04-polimorfismo/python/figura.py
"""
figura.py - CLASE ABSTRACTA

  abstract class Figura      ->  class Figura(ABC):
  abstract double area();    ->  @abstractmethod
                                 def area(self): ...

ABC = Abstract Base Class. Hay que importarla del modulo abc de la
libreria estandar: en Java "abstract" es una palabra del lenguaje,
en Python es una herramienta que se importa.
"""

from abc import ABC, abstractmethod


class Figura(ABC):

    def __init__(self, nombre):
        self._nombre = nombre

    @abstractmethod
    def area(self):
        """Cada hija esta OBLIGADA a implementarlo."""

    @abstractmethod
    def perimetro(self):
        """Idem."""

    def get_nombre(self):        # metodo normal, compartido por todas
        return self._nombre

File: 04-polimorfismo/java/Figura.java
/*
 * Figura.java - CLASE ABSTRACTA
 *
 * No se puede instanciar: "new Figura()" es error de compilacion.
 * Define QUE debe saber hacer toda figura, sin decir COMO.
 *
 * Es un archivo que describe una idea, no una cosa concreta.
 */
public abstract class Figura {

    protected String nombre;

    public Figura(String nombre) {
        this.nombre = nombre;
    }

    // METODOS ABSTRACTOS: sin cuerpo. Cada hijo esta OBLIGADO a implementarlos.
    public abstract double area();
    public abstract double perimetro();

    // Una clase abstracta SI puede tener metodos normales, compartidos por todos
    public String getNombre() {
        return this.nombre;
    }
}
